### ISSUE TITLE
Issue with 'onMapIsReady'

### ISSUE THREAD

**Issue by suvindran3** [liodali/osm_flutter#214]
Map initialization with user location seems to be not working so I used <strong>baseController.currentLocation()</strong> inside the method <strong>onMapIsReady</strong> using the <strong>mapState</strong> parameter like if <strong>mapState</strong> is true proceed with the <strong>currentLocation()</strong>. So everything was working perfectly fine till yesterday, Today the <strong>mapState</strong> is not getting true at all and I'm not getting any error too. I'm confused and have no idea what exactly is the issue here.

**Comment by liodali** [comment]
can you share some example of your widget to understand your use case more ?

**Comment by suvindran3** [comment]
Yeah here's the usage. Whole screen is used to display the map

<strong>OSMFlutter(
            controller: controller,
            trackMyPosition: true,
            initZoom: 17,
            onMapIsReady: (mapState) async {
              if (mapState) {
                await controller.osmBaseController.currentLocation();
                slideController.forward();
              }
            },
            androidHotReloadSupport: true,
          ), // OSMFlutter </strong>

and the controller,

<strong>final BaseMapController controller = MapController();</strong>

**Comment by liodali** [comment]
ok I will check that , i will give u my feedback

**Comment by liodali** [comment]
just notice i don't recommand you to use `BaseMapController` as type for the controller, use `MapController`
because with MapController type you don't need to access to the `osmBaseController` attribute to manipulate the map
you can only use it if you want to make you own custom controller with limited apis access

**Comment by liodali** [comment]
and also, i will fix problem related to trackMyPosition 
so in the future fix you don't need to call `currentLocation` in mapIsReady
and i have small question, is that problem related when application show you dialog to grant permission 
or something else ?

**Comment by suvindran3** [comment]
Okay so I changed the type to <code>MapController</code> thank you for that and no there was no issue with granting permission. The thing here is everything was working fine, I didn't change anything and today when I opened the map it just kept on loading. I tried drawing a road it worked just fine and tried <code>goToLocation</code> it worked too. So its just that <code>mapState</code>

**Comment by liodali** [comment]
check the new version `0.28.2`

**Comment by suvindran3** [comment]
<code>currentLocation()</code> is now working properly without explicitly calling it but the <code>mapState</code> is still not working. Thank you for responding and fixing this real quick.

**Comment by liodali** [comment]
try to use this in initState if your widget is statefullWidget
```dart
controller.listenerMapIsReady.addListener(()async{
if(controller.listenerMapIsReady.value){
/// put your logic
}

})

```

**Comment by suvindran3** [comment]
Okay I implemented what you've mentioned here and it worked totally fine but that method seems to have deprecated so tried the alternate method suggested there which is,

<pre>class YourOwnStateWidget extends State<YourWidget> with OSMMixinObserver {

   //etc
  @override
  void initState() {
    super.initState();
    controller.addObserver(this);
  }
    @override
    Future<void> mapIsReady(bool isReady) async {
      if (isReady) {
        /// put you logic
      }
    }
  @override
  Future<void> mapRestored() async {
    super.mapRestored();
    /// TODO
  }
    //etc
}</pre>

but this one didn't work.

**Comment by liodali** [comment]
still facing this issue ?

**Comment by suvindran3** [comment]
Yes and I'm facing new issues each day like today <code>drawRoad()</code> function didn't work properly like the first time I used, road was in blue color and the map view was animated to the location of the road drawn but now the road is in green thin line and the view is not changing to the road location. Is there any solution for this?

**Comment by liodali** [comment]
is there any log error 
or at least provide some config that you use it to drawRoad

**Comment by suvindran3** [comment]
This is the code I've written, I have a list of coordinates so I use first coordinate as <code>start</code> and last one as <code>end</code> and for intersection I use the remaining coordinates in that list.

<pre>
final RoadInfo roadInfo = await controller.drawRoad(
      GeoPoint(latitude: widget.startLat, longitude: widget.startLng),
      GeoPoint(latitude: widget.endLat, longitude: widget.endLng),
      intersectPoint: List.generate(
        widget.intersections.length - 2,
        (index) => GeoPoint(
          latitude:
              widget.intersections.elementAt(index + 1)[fieldName.lat.name],
          longitude:
              widget.intersections.elementAt(index + 1)[fieldName.lng.name],
        ),
      ),
    );
</pre>

and this is the log I got back and there is not even an error,  

<pre>D/BONUSPACK( 9340): OSRMRoadManager.getRoads:https://routing.openstreetmap.de/routed-car/route/v1/driving/80.2232118000,12.9881103000;80.2227806000,12.9882720000;80.2224673000,12.9886170000;80.2221616000,12.9889565000;80.2218088000,12.9892598000;80.2214064000,12.9894843000;80.2210756000,12.9898211000;80.2207216000,12.9901107000;80.2204301000,12.9904622000;80.2201284000,12.9908143000;80.2198421000,12.9911742000;80.2195361000,12.9915916000;80.2192750000,12.9919895000;80.2189962000,12.9924032000;80.2187306000,12.9927828000;80.2184537000,12.9931477000;80.2181376000,12.9935816000;80.2178573000,12.9939767000;80.2175656000,12.9943526000;80.2172968000,12.9947506000;80.2170455000,12.9951341000;80.2167976000,12.9956039000;80.2166479000,12.9961092000;80.2165615000,12.9966291000;80.2165390000,12.9970866000;80.2162633000,12.9974515000;80.2159317000,12.9978317000;80.2156083000,12.9982187000;80.2152930000,12.9985902000;80.2149476000,12.9989970000;80.2145938000,12.9993983000;80.2142753000,12.9997476000;80.2141439000,12.9998910000?alternatives=false&overview=full&steps=true
I/zygote  ( 9340): Background concurrent copying GC freed 31526(2MB) AllocSpace objects, 64(1404KB) LOS objects, 49% free, 3MB/6MB, paused 276us total 100.078ms
D/BONUSPACK( 9340): OSRMRoadManager.getRoads - finished
</pre>

**Comment by liodali** [comment]
in the first time you didn't test it like this?

**Comment by suvindran3** [comment]
First time I tested out with 4 sample coordinates like two intersection and It worked perfectly well, I tried that same sample now to check if the problem is in my side but turned out it didn't work so I'm kinda really confused right now

**Comment by liodali** [comment]
from where did you get the interesection points ? did you using api or any other routing service ?

**Comment by suvindran3** [comment]
Okay I found the issue. I missed the <code>roadOption</code> parameter in <code>drawRoad()</code>. That was a silly mistake, you've been really helpful and got to thank you for responding. Thank you :)

**Comment by suvindran3** [comment]
and btw I'm getting the coordinates from live tracking and each coordinate is 100 meters apart and I'm plotting them on the map.
Actually off topic, I got this one issue where I have to figure out if one set of coordinates is intersecting/overlapping another set like two roads will be drawn with those two sets of coordinates and got to find if they're intersecting/overlapping by code, we can see it easily but I need in terms of coding. You have any solution/thought on this?

**Comment by liodali** [comment]
those coordinates  you need to finetune them  to simple route (transform them to route with any routing api)
and after that you can compare them and find any intersection between them
see also this project called valhalla (https://github.com/valhalla/valhalla),
i will make more research on this

**Comment by suvindran3** [comment]
okay so I checked Valhalla it was helpful and I ended up finding the solution. Google have this geometry library for map functions and there's this one function called <code>isLocationOnPath</code> which checks if the given coordinate lies on a list of coordinates(polyline).

**Comment by suvindran3** [comment]
In OSMFlutter, Can't we draw more than one road using <code>controller.drawRoad()</code> at once? like when I draw the second road it's automatically removing the previous road.

**Comment by liodali** [comment]
for now i started looking to add animated marker
so after add that feature and advance little bit on the web support , i will add this feature soon
but i will create new issue for this feature that will be avalaible after 0.31+

**Comment by suvindran3** [comment]
Yeah got that! Thanks again without this plugin I'd be struggling with my current project

**Comment by liodali** [comment]
but for this issue still facing the same problem ?

**Comment by suvindran3** [comment]
Yeah this issue is still there. I even checked that today and another thing is that <code>mapIsLoading</code> is not working too. If I add any widget to <code>mapIsLoading</code> the map is not showing at all meaning its loading forever

**Comment by liodali** [comment]
:cry: i will checked, retry to create another empty project and add osm library to see the problem

**Comment by suvindran3** [comment]
I'll try that and let you know and one more thing. Is there any other way to highlight roads like an alternate to draw road? I need to highlight the covered roads by users

**Comment by liodali** [comment]
you want for example draw road with alternate road also ?

**Comment by suvindran3** [comment]
No not that. So I have this <code>List<List<GeoPoints<n>>></code> which are the covered roads of different users. I have to loop through the list and draw a road for each <code>List<GeoPoints<n>></code> where I use the first and last GeoPoint as <code>start</code> and <code>end</code> and the remaining GeoPoints as the <code>intersections</code>. Now I thought I'll just draw multiple roads to highlight the covered roads but multiple roads are not supported now. Is there any other way that I can execute this like I have to somehow highlight the covered roads?

**Comment by suvindran3** [comment]
Thank you for taking your time to respond and for the assistance you've been providing with this matter.
